src/audio/tdfb/tdfb_direction.c

```diff
@@ -0,0 +1,559 @@
+		d = cd->mic_locations[i + 2].x - cd->mic_locations[i].x;
+		e = cd->mic_locations[i + 2].y - cd->mic_locations[i].y;
+		f = cd->mic_locations[i + 2].z - cd->mic_locations[i].z;
+		cross_product_s16(&px, &py, &pz, a, b, c, d, e, f);
+
+		/* Allow small room for rounding error with |x| > 1 in check */
+		if (ABS(px) > 1 || ABS(py) > 1 || ABS(pz) > 1)
+			return false;
+	}
+
+	return true;
+}
+
+int tdfb_direction_init(struct tdfb_comp_data *cd, int32_t fs, int ch_count)
+{
+	struct sof_eq_iir_header_df2t *filt;
+	int64_t *delay;
+	int32_t d_max;
+	int32_t t_max;
+	size_t size;
+	int n;
+	int i;
+
+	/* Select emphasis response per sample rate */
+	switch (fs) {
+	case 16000:
+		filt = (struct sof_eq_iir_header_df2t *)iir_emphasis_16k;
+		break;
+	case 48000:
+		filt = (struct sof_eq_iir_header_df2t *)iir_emphasis_48k;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	/* Allocate delay lines for IIR filters and initialize them */
+	size = ch_count * iir_delay_size_df2t(filt);
+	delay = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, size);
+	if (!delay)
+		return -ENOMEM;
+
+	cd->direction.df2t_delay = delay;
+	for (i = 0; i < ch_count; i++) {
+		iir_init_coef_df2t(&cd->direction.emphasis[i], filt);
+		iir_init_delay_df2t(&cd->direction.emphasis[i], &delay);
+	}
+
+	/* Unit delay length in Q1.31 seconds */
+	cd->direction.unit_delay = Q_SHIFT_LEFT(1LL, 0, 31) / fs;
+
+	/* Get max possible mic-mic distance, then from max distance to max time t = d_max / v,
+	 * t is Q1.15, d is Q4.12, v is Q9.0
+	 */
+	d_max = max_mic_distance(cd);
+	t_max = Q_SHIFT_LEFT(d_max, 12, 15) / SPEED_OF_SOUND;
+
+	/* Calculate max lag to search and allocate delay line for cross correlation
+	 * compute. Add one to make sure max possible lag is in search window.
+	 */
+	cd->direction.max_lag = Q_MULTSR_32X32(fs, t_max, 0, 15, 0) + 1;
+	n = (cd->max_frames + (2 * cd->direction.max_lag + 1)) * ch_count;
+	cd->direction.d_size =  n * sizeof(int16_t);
+	cd->direction.d = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, cd->direction.d_size);
+	if (!cd->direction.d)
+		goto err_free_iir;
+
+	/* Set needed pointers to xcorr delay line, advance write pointer by max_lag to keep read
+	 * always behind write
+	 */
+	cd->direction.d_end = cd->direction.d + n;
+	cd->direction.rp = cd->direction.d;
+	cd->direction.wp = cd->direction.d + ch_count * (cd->direction.max_lag + 1);
+
+	/* xcorr result is temporary but too large for stack so it is allocated here */
+	cd->direction.r_size = (2 * cd->direction.max_lag + 1) * sizeof(int32_t);
+	cd->direction.r = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, cd->direction.r_size);
+	if (!cd->direction.r)
+		goto err_free_all;
+
+	/* Check for line array mode */
+	cd->direction.line_array = line_array_mode_check(cd);
+
+	/* Initialize direction to zero radians, set initial step sign to +1 */
+	cd->direction.az = 0;
+	cd->direction.step_sign = 1;
+	return 0;
+
+err_free_all:
+	rfree(cd->direction.d);
```

Comment:
if we know these are NULL initially then we don't need 2 labels

Reply:
Sorry, I'm missing something. There's 3 successive rzalloc() calls. if 3rd fails 1-2 are freed, etc.

Reply:
I think this looks ok for the 3 allocs unless I'm missing something ?

Reply:
just a minor optimization suggestion, nothing major, we could reduce the labels and rely on rfree(NULL) being a no-op